module/apmpgx/tracer.go

```diff
@@ -0,0 +1,185 @@
+// not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package apmpgx // import "go.elastic.co/apm/module/apmpgx/v2"
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"strings"
+	"time"
+
+	"go.elastic.co/apm/module/apmsql/v2"
+	"go.elastic.co/apm/v2"
+
+	"github.com/jackc/pgx/v4"
+)
+
+const (
+	//querySpanType is setting action for query expression trace in APM server.
+	querySpanType = "db.postgresql.query"
+
+	//querySpanType is setting action for copy expression trace in APM server.
+	copySpanType = "db.postgresql.copy"
+
+	//querySpanType is setting action for batch expression trace in APM server.
+	batchSpanType = "db.postgresql.batch"
+)
+
+var (
+	// ErrUnsupportedPgxVersion is indicating that data doesn't contain value for "time" key. This field appeared in pgx v4.17
+	ErrUnsupportedPgxVersion = errors.New("this version of pgx is unsupported, please upgrade to v4.17+")
+
+	// ErrInvalidType is indicating that field type doesn't meet expected one
+	ErrInvalidType = errors.New("invalid field type")
+)
+
+// tracer struct contains pgx.ConnConfig inside and pgx.Logger implementation.
+type tracer struct {
+	// cfg is the pgx.ConnConfig which used for setting metadata in spans such as host, port, etc.
+	cfg *pgx.ConnConfig
+
+	// logger used for writing data to log. If it's nil, then data won't be written to log, and only spans will be created.
+	logger pgx.Logger
+}
+
+// Instrument is getting pgx.ConnConfig and wrap logger into tracer.
+// It's safe to pass nil logger into pgx.ConnConfig, if so, then only spans will be created
+func Instrument(cfg *pgx.ConnConfig) {
+	cfg.Logger = &tracer{
+		cfg:    cfg,
+		logger: cfg.Logger,
+	}
+}
+
+// Log is getting type of SQL expression from msg and run suitable trace.
+// If logger in tracer struct isn't nil, than log will be
+// written to your logger that implements pgx.Logger interface.
+func (t *tracer) Log(ctx context.Context, level pgx.LogLevel, msg string, data map[string]interface{}) {
+	if t.logger != nil {
+		t.logger.Log(ctx, level, msg, data)
+	}
+
+	switch msg {
+	case "Query", "Exec":
+		t.QueryTrace(ctx, data)
+	case "CopyFrom":
+		t.CopyTrace(ctx, data)
+	case "SendBatch":
+		t.BatchTrace(ctx, data)
+	}
+}
+
+// QueryTrace traces query and creates spans for them.
+func (t *tracer) QueryTrace(ctx context.Context, data map[string]interface{}) {
+	statement, ok := data["sql"].(string)
+	if !ok {
+		apm.CaptureError(ctx,
+			fmt.Errorf("%w: expect string, got: %T",
+				ErrInvalidType,
+				data["sql"],
+			),
+		).Send()
+		return
+	}
```

Comment:
@axw one more question - should I capture and send type casting errors into APM? I didn't find something like this in other modules, but I think it will be good for understanding that something inside module (or its dependency) is broken

Reply:
There's no precedent and I'm a bit on the fence about it, but I think it's fine to keep it for now at least. There's not really a good alternative at the moment. Generally errors are about the service, rather than about the instrumentation. I suppose ideally there would be some way to log errors about the instrumentation - we can revise it later.